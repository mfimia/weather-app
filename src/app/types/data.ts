export interface IWeather {
  date: string
  temperature: number
  condition: string
  iconUrl: string
  humidity?: number
  windSpeed?: number
}

export interface UIComponent {
  type: string
  text: string
}

export interface CityWeatherData {
  [cityName: string]: IWeather[]
}

export interface UIData {
  ui: UIComponent[]
}

export type IWeatherData = CityWeatherData & UIData

export interface IWeatherPayload {
  cityName: Cities
  cityData: IWeather[]
  ui: UIComponent[]
}

export enum Cities {
  Düsseldorf = 'Düsseldorf',
  Málaga = 'Málaga',
  Berlin = 'Berlin'
}
