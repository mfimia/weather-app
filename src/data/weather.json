{
  "Düsseldorf": [
    {
      "date": "current",
      "temperature": 15,
      "condition": "Cloudy",
      "humidity": 68,
      "windSpeed": 10,
      "iconUrl": "/cloudy.svg"
    },
    {
      "date": "Monday",
      "temperature": 14,
      "condition": "Cloudy",
      "iconUrl": "/cloudy.svg"
    },
    {
      "date": "Tuesday",
      "temperature": 15,
      "condition": "Partly Cloudy",
      "iconUrl": "/partly_cloudy.svg"
    },
    {
      "date": "Wednesday",
      "temperature": 16,
      "condition": "Sunny",
      "iconUrl": "/sunny.svg"
    }
  ],
  "Málaga": [
    {
      "date": "current",
      "temperature": 25,
      "condition": "Sunny",
      "humidity": 50,
      "windSpeed": 5,
      "iconUrl": "/sunny.svg"
    },
    {
      "date": "Monday",
      "temperature": 26,
      "condition": "Partly Cloudy",
      "iconUrl": "/partly_cloudy.svg"
    },
    {
      "date": "Tuesday",
      "temperature": 25,
      "condition": "Sunny",
      "iconUrl": "/sunny.svg"
    },
    {
      "date": "Wednesday",
      "temperature": 24,
      "condition": "Cloudy",
      "iconUrl": "/cloudy.svg"
    }
  ],
  "Berlin": [
    {
      "date": "current",
      "temperature": 13,
      "condition": "Rainy",
      "humidity": 75,
      "windSpeed": 12,
      "iconUrl": "/rainy.svg"
    },
    {
      "date": "Monday",
      "temperature": 12,
      "condition": "Rainy",
      "iconUrl": "/rainy.svg"
    },
    {
      "date": "Tuesday",
      "temperature": 13,
      "condition": "Cloudy",
      "iconUrl": "/cloudy.svg"
    },
    {
      "date": "Wednesday",
      "temperature": 14,
      "condition": "Partly Cloudy",
      "iconUrl": "/partly_cloudy.svg"
    }
  ],
  "ui": [
    {
      "type": "header",
      "text": "Weather for {cityName}"
    },
    {
      "type": "weatherInfo"
    },
    {
      "type": "daysForecast"
    }
  ]
}