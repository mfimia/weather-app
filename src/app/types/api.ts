import { IWeatherPayload } from './data'

export interface IApiResponse {
  code: number
  data: IWeatherPayload
}
