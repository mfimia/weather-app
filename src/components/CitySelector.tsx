import { Cities } from '@/app/types/data'
import { FC, Dispatch, SetStateAction } from 'react'

interface ICitySelectorProps {
  city: Cities
  setCity: Dispatch<SetStateAction<Cities>>
}

const CitySelector: FC<ICitySelectorProps> = ({ city, setCity }) => {
  return (
    <div className="flex justify-between items-center mb-5">
      <select
        value={city}
        onChange={(e) => setCity(e.target.value as Cities)}
        className="p-2 border rounded"
      >
        {Object.values(Cities).map(cityName => <option key={cityName} value={cityName}>{cityName}</option>)}
      </select>
    </div>
  )
}

export default CitySelector