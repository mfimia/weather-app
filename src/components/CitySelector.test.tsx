import { render, screen, fireEvent } from '@testing-library/react'
import CitySelector from './CitySelector'
import { Cities } from '../app/types/data'

describe('CitySelector', () => {
  let mockCity: Cities
  let mockSetter: jest.Mock
  let selectElement: HTMLElement

  beforeEach(() => {
    mockCity = Cities.Berlin
    mockSetter = jest.fn()

    render(<CitySelector city={mockCity} setCity={mockSetter} />)

    selectElement = screen.getByRole('combobox')
  })

  it('renders a SELECT element with 3 options', () => {
    const { container } = render(
      <CitySelector city={mockCity} setCity={mockSetter} />
    )

    expect(selectElement).toBeVisible()
    expect(container).toMatchSnapshot()
  })

  it('should display all the cities in the Cities enum as select options', () => {
    Object.values(Cities).forEach((city) => {
      const option = screen.getByText(city)

      expect(option).toBeInTheDocument()
    })
  })

  it('should call the city setter when a new city is selected', () => {
    fireEvent.change(selectElement, { target: { value: Cities.Málaga } })

    expect(mockSetter).toHaveBeenCalledWith(Cities.Málaga)
  })
})
