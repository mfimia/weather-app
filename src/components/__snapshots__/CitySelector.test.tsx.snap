// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitySelector renders a SELECT element with 3 options 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between items-center mb-5"
  >
    <select
      class="p-2 border rounded cursor-pointer"
    >
      <option
        value="Düsseldorf"
      >
        Düsseldorf
      </option>
      <option
        value="Málaga"
      >
        Málaga
      </option>
      <option
        value="Berlin"
      >
        Berlin
      </option>
    </select>
  </div>
</DocumentFragment>
`;
